src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -0,0 +1,248 @@
+            if (!_triggerOperations.TryCreateFactory(_ruleOptions.Trigger.Type, out ICollectionRuleTriggerFactoryProxy factory))
+            {
+                throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_CouldNotMapToTrigger, _ruleOptions.Trigger.Type));
+            }
+
+            using CancellationTokenSource durationCancellationSource = new();
+            using CancellationTokenSource linkedCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(
+                durationCancellationSource.Token,
+                token);
+
+            CancellationToken linkedToken = linkedCancellationSource.Token;
+
+            TimeSpan? actionCountWindowDuration = _ruleOptions.Limits?.ActionCountSlidingWindowDuration;
+            int actionCountLimit = (_ruleOptions.Limits?.ActionCount).GetValueOrDefault(CollectionRuleLimitsOptionsDefaults.ActionCount);
+            Queue<DateTime> executionTimestamps = new(actionCountLimit);
+
+            // Start cancellation timer for graceful stop of the collection rule
+            // when the rule duration has been specified. Conditionally enable this
+            // based on if the rule has a duration limit.
+            TimeSpan? ruleDuration = _ruleOptions.Limits?.RuleDuration;
+            if (ruleDuration.HasValue)
+            {
+                durationCancellationSource.CancelAfter(ruleDuration.Value);
+            }
+
+            try
+            {
+                bool completePipeline = false;
+                while (!completePipeline)
+                {
+                    TaskCompletionSource<object> triggerSatisfiedSource =
+                        new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                    ICollectionRuleTrigger trigger = null;
+                    try
+                    {
+                        trigger = factory.Create(
+                            _endpointInfo,
+                            () => triggerSatisfiedSource.TrySetResult(null),
+                            _ruleOptions.Trigger.Settings);
+
+                        if (null == trigger)
+                        {
+                            throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, Strings.ErrorMessage_TriggerFactoryFailed, _ruleOptions.Trigger.Type));
+                        }
+
+                        // Start the trigger.
+                        await trigger.StartAsync(linkedToken).ConfigureAwait(false);
+
+                        // The pipeline signals that it has started just after starting a non-startup trigger.
+                        // Instances with startup triggers signal start after having finished executing the action list.
+                        if (trigger is not ICollectionRuleStartupTrigger)
+                        {
+                            // Signal that the pipeline trigger is initialized.
+                            _startedSource.TrySetResult(null);
```

Comment:
Do we have to special case startup triggers here btw? Perhaps their callback could return right away and satisfy the trigger condition?

Reply:
Yes, because a startup trigger signals that the runtime should be resumed after all actions have run (by default) and a non-startup trigger signals that the runtime should be resume just after the trigger has started. The signal that the runtime is able to be resumed is by completing startup of the pipeline so that (in a future change) the collection rule service will just say "start a pipeline instance for each rule on this process" and then immediately resume the runtime after each pipeline declares that it has started.